src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+---
+title: Patterns
+description: Summary of patterns in Dart.
+---
+
+{{site.alert.version-note}}
+  Patterns require a [language version][] of at least 3.0.
+{{site.alert.end}}
+
+Patterns are a syntactic category in the Dart language, like statements and expressions.
+They represent the form of a value, not an actual value.
+
+This page describes:
+- What patterns do (match and destructure)
+- Where patterns are allowed in Dart code
+- Common use cases for patterns
+
+To learn about the different types of patterns, visit the [pattern type reference][types].
+
+## Matching
+
+A pattern is always tested against a value to determine if the value has the form
+you expect, or in other words, if the value _matches_ the pattern. 
+
+What constitutes a match depends on the [pattern type][types].
+For example, matching a constant pattern means the value is equivalent to the pattern:
+
+```dart
+switch (number) {
+  case 1: ...        // Constant pattern 1 matches if number == 1
+}
+```
+
+Many patterns make use of subpatterns (sometimes called outer and inner patterns,
+respectively). Patterns match recursively on their subpatterns.
+For example, the individual fields of any [collection-type][] pattern could be 
+[variable patterns][variable] or [constant patterns][constant]:
+
+```dart
+switch (obj) {       // List pattern [a,b] matches obj first if obj
+  case [a, b]: ...   // is a list wth two fields, then if its  
+}                    // fields match the constant subpatterns a and b  
+```
+
+Patterns can also _partially_ match, by using a [wildcard pattern][] as a place holder,
+or a [rest element][] in the case of list patterns.
+
+If you need to ensure `null` values aren't silently treated as match failures,
+use the [null-check pattern][] while matching.
+
+## Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
+print(a + b + c);          // numList into individual new parts. Prints 6.
+```
+
+You can use the [null-check pattern][] to check for `null` while destructuring. 
+
+All patterns attempt to match the objects they're compared against, but not all
```

Comment:
I'm not sure about this paragraph.  What precisely do you want to the reader to take away here?  It's true that whether a pattern destructures or not depends on the type of pattern, not sure about the context.  Maybe there's some confusion around destructure in the sense of "destructuring assignment" vs "extract into parts"?  That is, the following pattern destructures (extracts) a list into parts, whether it is used is in a destructuring assignment, or in a switch:

```
  var [a, b] = someList; // Destructuring assignment: unconditionally extract the parts from something that must be a list.
  switch (someList) {
     case [var a, var b]: // Switch case: extract the parts from something that might be a list, if it is a list (and otherwise fall through to the next case).
  }
```

Reply:
I think I somewhat misunderstood the point of this and the previous section, which perhaps partially reflects me being too close to the feature, and maybe reflects me reading to quickly, but might also suggest that other people might also go astray?  I think what you are trying to say with these sections is roughly "Hey, in general, there are two things going on in patterns.  First, given a value, you might want to check whether the value has a certain shape, or is a certain constant, or is equal to something else, or has a certain type.  And second, given a value, you might want a convenient declarative syntax to break it into a bunch of its constituent parts, possibly binding variables to some or all of those parts in the process.  The first we call matching, and the second we call destructuring.  Patterns let us to both at once. Depending on the context and the shape of the pattern, a pattern may match a value, destructure a value, or both.  Here are some examples to show you what we mean by matching, and here are some to show you what we mean by destructuring".  Is that a fair (rough summary)?  If so, I think these sections are actually pretty close, I just think that maybe a little bit of introductory text like the above to give the reader a roadmap of what's going on might avoid sending people like me down the wrong road (I was kind of reading this as talking more about refutable vs irrefutable contexts, whereas I think you mean something more like what I describe above).  Does that make sense?

Reply:
Yes that's it exactly. I added a new section header to summarize this, and then made the matching and destructuring sections subsections under that introduction